geo_tag/geo_tag.py

```diff
@@ -0,0 +1,221 @@
+            geo_tag["countyID"] = geo_content[1]
+            geo_tag["countyName"] = geo_content[3]
+            geo_tag["cityID"] = geo_content[0]
+            geo_tag["cityName"] = city_state_name.split(",")[0]
+            geo_tag["coordinate"] = coord
+            geo_tag["source"] = flag
+
+            tweet_json['geo_tag'] = geo_tag
+            print(geo_tag)
+            
+        else:
+            raise ValueError("Not find county.")
+
+    def __infer_geo_from_place(self, tweet_json: dict, coord: tuple):
+        place = tweet_json.get('place')
+        if place is not None:
+            # extract the city and state abbrev
+            city_state_name = place.get('full_name')
+            if city_state_name is None:
+                try:
+                    raise KeyError()
+                except KeyError:
+                    print("Not find full_name key in place.")
+                return
+
+            self.__extract_geo_tag_from_city_and_state(tweet_json, coord, city_state_name, "place")
+
+    def __infer_geo_from_coord(self, tweet_json: dict, coord: tuple):
+        # geopy require (Lat, Long)
+        t = (coord[1], coord[0])
+        location = self.geo_locator.reverse(t)
+
+        if location is None:
+            tweet_json['geo_tag'] = {"coordinate": coord}
+            return
+
+        # make up the (city, state) as key to search
+        target_key = location.address.split(",")[3].strip() + ', ' + location.address.split(",")[5].strip()
+        geo_content = self.cache.get(target_key)
+
+        if geo_content is not None:
+            geo_tag = dict()
+            geo_tag["stateID"] = geo_content[2]
+            geo_tag["stateName"] = location.address.split(",")[5].strip()
+            geo_tag["countyID"] = geo_content[1]
+            geo_tag["countyName"] = geo_content[3]
+            geo_tag["cityID"] = geo_content[0]
+            geo_tag["cityName"] = location.address.split(",")[3].strip()
+            geo_tag["coordinate"] = coord
+            geo_tag["source"] = "coordinate"
+
+            tweet_json['geo_tag'] = geo_tag
+            print(geo_tag)
+
+        else:
+            raise ValueError("Not find county.")
+
+    def __infer_geo_from_user(self, tweet_json, coord: tuple):
+        user = tweet_json.get('user')
+        if user is not None:
+            # extract the city and state abbrev
+            city_state_name = user.get('location')
+            if city_state_name is None:
+                try:
+                    raise KeyError()
+                except KeyError:
+                    print("Not find full_name key in place.")
+                return
+
+            self.__extract_geo_tag_from_city_and_state(tweet_json, coord, city_state_name, "user")
+        else:
+            # The methods above are all not working, for now we skip.
+            # TODO: Use NLP to infer geo_tag from text in tweet.
+            pass
+
+    def tag_one_tweet(self, tweet_json: dict, mode: int):
```

Comment:
this is the main entry API. it should handle any exceptions found in helper functions, log them out and tag tweet in a default mode: still have geo_tag attribute, but values are all `None`.

Reply:
OK, if we log out all the errors, there might be so many errors log in the terminal.

Reply:
we log them into files. you can see other modules in this repo how to configure file logger.